File: backend/src/NotSpotify.Api/Models/Repost.cs
namespace NotSpotify.Api.Models;

/// <summary>
/// A repost (re-share) by a user to their followers.
/// Extends the asymmetric follow graph — followers see reposts in their feed.
/// Polymorphic: can reference a Track, Album, or Playlist.
/// </summary>
public class Repost
{
    public Guid Id { get; set; } = Guid.NewGuid();

    /// <summary>Who reposted.</summary>
    public Guid UserId { get; set; }
    public ApplicationUser User { get; set; } = null!;

    public Guid? TrackId { get; set; }
    public Track? Track { get; set; }

    public Guid? AlbumId { get; set; }
    public Album? Album { get; set; }

    public Guid? PlaylistId { get; set; }
    public Playlist? Playlist { get; set; }

    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
}

File: frontend/src/types/repost.ts
import type { UserRef } from './user'
import type { Track } from './track'
import type { Album } from './album'
import type { Playlist } from './playlist'

export interface Repost {
  id: string
  user: UserRef
  trackId: string | null
  albumId: string | null
  playlistId: string | null
  track: Track | null
  album: Album | null
  playlist: Playlist | null
  createdAt: string
}
